### PR TITLE
Make astropy.utils.decorators.wraps much more like functools.wraps

### PR DESCRIPTION
**Pull request by Cadair**
Here is a controversial fix to #7551. This removes two features of `astropy.utils.decorators.wraps` the `exclude_args` method and the way it overwrote the functions name.

In return for this, we make use of the new `__signature__` attribute in Python 3 and we probably don't need a lot more code that I have currently deleted in here.

It seems that the only place astropy was using `exclude_args` was in the test runner, which is now irrelevant as my changes to that mean it generates the documentation automatically anyway. And my opinion on the name thing is that it's deeply horrific and probably never should have been done in the first place :wink:

### CODE REVIEW

**Comment by Cadair** [comment]
It's also worth noting, that there is only one other use of `make_function_with_signature` in astropy, in modeling. Which could probably be replaced with direct manipulation of the Python 3 signature machinery. So we could probably remove all of that as well.

**Comment by astrofrog** [comment]
Ok so just to be clear, ``wraps`` in Python 3.5 now no longer requires us to have our own implementation? If so, that's fine by me in principle

File: astropy/modeling/core.py [hunk 1]
```diff
@@ -332,7 +332,8 @@ def __call__(self, **kwargs):
 
             kwargs.append((param.name, param.default))
 
-        __call__ = make_function_with_signature(__call__, ('self',), kwargs)
+        # __call__ = make_function_with_signature(__call__, ('self',), kwargs)
```
Comment (bsipocz) [inline astropy/modeling/core.py]: don't comment but remove line

**Comment by Cadair** [comment]
@astrofrog Yes, but not quite with perfect feature parity.

**Comment by pllim** [comment]
> I am not going to waste my time debugging it

Looks like a "catch 22" situation. I am not convinced until tests can pass. :wink:

**Comment by Cadair** [comment]
haha, the question is, are we willing to loose the `exclude_args` and name trickery for the sake of this refactor? If so then I will try and make it work. Obviously if I can't make it work it's just a waste of everyones time...

**Comment by pllim** [comment]
I cannot comment on the `nddata` stuff but maybe @mwcraig or @MSeifert04 can?

**Comment by mwcraig** [comment]
I’m gong to be offline for the next week, can look when I’m back

**Comment by mhvk** [comment]
Very much in favour from the peanut gallery! Though since it is actually in the documentation, I think `wraps` also would need to be deprecated first (as far as I'm concerned, this can be by returning the python standard version with a deprecation warning that it will not stay in `astropy.utils.decorators`).

**Comment by bsipocz** [comment]
I didn't find a single place on github where this was used, so we most probably can get away with @mhvk's suggestion for a very quick deprecation period.

**Comment by Cadair** [comment]
@bsipocz did you also check for `make_function_with_signature` by chance? I assume we should deprecate that for a little while as well?

**Comment by bsipocz** [comment]
@Cadair - There is minimal use for that, and most of the hits seems to be in copies of astropy itself: https://github.com/search?q=make_function_with_signature&type=Code

I've only found  one genuine use of it in one of Perry's repos, so my guess is the usage is indeed fairly low.

**Comment by MSeifert04** [comment]
I like it provided we don't completely break `support_nddata` :)

**Comment by pllim** [comment]
Too controversial to merge so close to freeze time, so re-milestoned. 😬

**Comment by Cadair** [comment]
I think this is one for the meeting in Dec.

**Comment by pllim** [comment]
> meeting in Dec

OK, added to breakout planning board.

**Comment by mhvk** [comment]
Still in favour of this.

**Comment by pllim** [comment]
@mhvk , there is a "keep open" label if you want to use it, FYI

**Comment by bsipocz** [comment]
@Cadair  - Could you please rather rebase this that keeping it alive by the label? IMO we shouldn't over use the label otherwise easily end up it the previous situation of having long abandoned PRs all over.

**Comment by mhvk** [comment]
@bsipocz - that was the fault of me and @pllim - but point taken that a rebase would be better.

**Comment by Cadair** [comment]
rebased.

**Comment by mhvk** [comment]
OK, the next step obviously is to make sure `support_nddata` keeps working...

File: astropy/utils/tests/test_codegen.py [hunk 2]
```diff
@@ -36,6 +36,7 @@ def crashy_function(*args, **kwargs):
         # make_function_with_signature call was one
         tb_lines = traceback.format_tb(tb)
         assert '1 / 0' in tb_lines[-1]
-        assert line in tb_lines[-2] and 'line =' not in tb_lines[-2]
+        # The deprecated decorator breaks this check
+        # assert line in tb_lines[-2] and 'line =' not in tb_lines[-2]
```
Comment (bsipocz) [inline astropy/utils/tests/test_codegen.py]: any chance to fix this?

Comment (Cadair) [reply astropy/utils/tests/test_codegen.py]: I tried to work out what was causing it and given the function it is testing is now deprecated, I didn't think it was worth the fight.

Comment (bsipocz) [reply astropy/utils/tests/test_codegen.py]: then remove it?

**Comment by mwcraig** [comment]
@bsipocz @Cadair -- I'm working on getting the last couple bits here fixed.

File: astropy/utils/tests/test_codegen.py (hunk 2 above)
Comment (mwcraig) [reply astropy/utils/tests/test_codegen.py]: Fixed this...

**Comment by mwcraig** [comment]
Waiting to see if doc issue still occurs. It does not locally (though I'm seeing an error in an unrelated part of the docs)

**Comment by pllim** [comment]
Doc still failed. FYI

```
Warning, treated as error:
toctree references unknown document 'api/astropy.utils.decorators.wraps'
Sphinx Documentation subprocess failed with return code 2
Exited with code 2
```

**Comment by mwcraig** [comment]
> Doc still failed. FYI

Yeah, still can't reproduce locally, which is making it hard to troubleshoot....

**Comment by mwcraig** [comment]
OK, reproduced and fixed (hopefully). The other changes in the last push were to replace imports of `astropy.utils.wraps` with `functools.wraps` so that we can eventually not provide `wraps` at all in `astropy.utils`.

**Comment by Cadair** [comment]
Awesome!! Thanks @mwcraig

**Comment by bsipocz** [comment]
Since the deprecation from this PR has been reverted in #8624, I'm adding the `no-changelog-entry-needed` label to stop the release script from complaining.

**Comment by bsipocz** [comment]
For the sake of the exercise I'm looking at recent benchmark regressions and weirdly this came up as a reason for a 1.3-1.5x slowdown in the convolution benchmarks.
http://www.astropy.org/astropy-benchmarks/#regressions?sort=1&dir=desc

**Comment by pllim** [comment]
Hmm... Do you think this is because it is using `modeling` and `modeling` is using the deprecated code?

https://github.com/astropy/astropy/blob/7a81d70882d9797859954134489a5c802a067606/astropy/modeling/core.py#L385

**Comment by bsipocz** [comment]
Yes, I think it's modeling related, but the regression should have show up in the deprecating PR if it's because of that. I suspect just there are some overheads here that we didn't think about, and/or some non optimal usage of modeling in convolution.

I would say that we need to keep an eye on those benchmarks after the refactoring PR is merged in. Or, do you @astrofrog think  that we can  run the benchmarks on the PR to check  the performance implications? (though either way the refactoring will be merge, so maybe  it's not urgent to check it in the PR stage).

**Pull request merged**